airbyte-cdk/python/airbyte_cdk/sources/availability_strategy.py

```diff
@@ -0,0 +1,145 @@
+from requests import HTTPError
+from typing import Dict, List, Optional, Text, Tuple
+
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources import Source
+
+
+class AvailabilityStrategy(ABC):
+    """
+    Abstract base class for checking stream availability
+    """
+
+    @abstractmethod
+    def check_availability(self, source: Source, logger: logging.Logger, stream: Stream) -> Tuple[bool, any]:
+        """
+        Checks stream availability.
+
+        :param source: source
+        :param logger: source logger
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+
+
+class HTTPAvailabilityStrategy(AvailabilityStrategy):
+    def check_availability(self, source: Source, logger: logging.Logger, stream: Stream) -> Tuple[bool, str]:
+        """
+        Check stream availability by attempting to read the first record of the
+        stream.
+
+        :param source: source
+        :param logger: source logger
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+        try:
+            # Some streams need a stream slice to read records (e.g. if they have a SubstreamSlicer)
+            stream_slice = self._get_stream_slice(stream)
+            records = stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice)
+            next(records)
+        except HTTPError as error:
+            return self.handle_http_error(source, logger, stream, error)
+        return True, None
+
+    def handle_http_error(self, source: Source, logger: logging.Logger, stream: Stream, error: HTTPError):
+        """
+        Override this method to define error handling for various `HTTPError`s
+        that are raised while attempting to check a stream's availability.
+
+        :param source: source
+        :param logger: source logger
+        :param stream: stream
+        :param error: HTTPerror raised while checking stream's availability.
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+        if error.response.status_code == requests.codes.FORBIDDEN:
+            error_message = "This is most likely due to insufficient permissions on the credentials in use. "
+            error_message += self._visit_docs_message(source, logger)
+            return False, error_message
+
+        error_message = repr(error)
+        return False, error_message
```

Comment:
Similar to the other PR. Is it safer to default this to `True`? Otherwise ,we would be hiding lots of errors. Or maybe we want to skip ahead on 500. I would expect a failed sync.

Reply:
To answer both you and @alafanechere's same question, I had initially thought we wanted to except all HTTPErrors, since otherwise this would fix 403 issues but nothing else. I think the reason I thought this was that I was thinking that the main goal of this PR was to stop failing streams from failing a whole sync, when it's really more about gracefully handling the case of 403ing.

I've changed it to default to `True` (aka still available, don't catch errors, things will probably fail) for other HTTPErrors that might pop up. This leads us to more failed syncs (less than before, but more than if we excepted them all), but we can check in on which HTTPError is the new most common cause of stream sync failures and handle that code, and iterate that way. 

Alternatively we could except all HTTPErrors so that the "partial sync" (not yet a concept but you get the idea) still succeeds, and directly report the errors that are excepted so that we can still do that iteration, though its worth noting in this case that we're less motivated by errored syncs to fix those...

>  Or maybe we want to skip ahead on 500. I would expect a failed sync.

Could you elaborate on this? I'd definitely expect a failed sync on a 500, but that doesn't mean that all of the streams have to fail if their endpoints are still working. That's the main argument toward my initial implementation. 

Something that could cause false positives would be something like rate limiting and 429s, but we should already have code that handles that with retries -- one thing on my TODO is to make sure that works as expected, i.e. that the amount of retrying on the `next(records)` call in `check_connection` is the same as when reading records in the read function itself. We don't want to have to deal with retryable error code in this class, I think.

Reply:
> I had initially thought we wanted to except all HTTPErrors, since otherwise this would fix 403 issues but nothing else
I think, for now, only handling a specific set of HTTPError is safer: we know precisely that a stream is unavailable for a permission error and can support the user accordingly. 

I don't think transient HTTP errors like 429 or 503 should be used to characterize stream availability at the moment as these errors can disappear during the read.